Add Zephyr support for Mediatek MT8365 platform

andrew-mtk opened the pull request:
Add Zephyr support for Mediatek MT8365 platform. This MR depends on [https://github.com/zephyrproject-rtos/hal_xtensa/pull/36](https://github.com/zephyrproject-rtos/hal_xtensa/pull/36), so  `west.yml` has to be updated accordingly.

andrew-mtk committed:
soc: mediatek: adsp: Add MT8365 SOC

Add support for MediaTek MT8365 SoC. This includes adding basic
initialization code, IRQ handling, IPI driver, MPU configuration,
and necessary build system files.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
drivers: timer: Add support for MT8365 timer

Make changes to the timer driver to support MT8365 SoC.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk committed:
boards: mediatek: Add board support files for MT8365

Add board configuration files for MT8365 platform Audio DSP.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

nordicjm commented on boards/mediatek/mt8365/mt8365_adsp.dts:
```diff
@@ -0,0 +1,84 @@
+#include "afe-mt8365.dts"
+
+	}; /* soc */
+
+	chosen { };
+	aliases { };
```
```suggestion
```

nordicjm commented on soc/mediatek/mt8xxx/mt8365/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+
+config NUM_2ND_LEVEL_AGGREGATORS
+	default 1
+config 2ND_LVL_INTR_00_OFFSET
```
missing newline gap line 8

nordicjm commented on soc/mediatek/mt8xxx/mt8365/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if SOC_SERIES_MT8365
+
```
this PR is not properly structured, this commit adds soc files and 1 board file, fix it so there is 1 commit adding the soc files, 1 commit adding the board files

andyross commented on soc/mediatek/mt8xxx/irq.c:
```diff
@@ -35,6 +35,12 @@ static const struct device *irq_dev(unsigned int *irq_inout)
+	if ((*irq_inout & 0xff) == 1) {
+		*irq_inout = (*irq_inout >> 8) - 1;
+	}
+	return DEVICE_DT_GET(DT_INST(0, mediatek_adsp_intc));
```
Not a fault of this PR, but the hackery in this file (which really is a wart in Zephyr, we lack a generic IRQ routine generator even though DTS gives us all the information needed to generate it) is getting a little ugly.  Probably time to start thinking about a rework.

kartben reviewed (dismissed):
This is missing a documentation page

andrew-mtk committed:
boards: mediatek: Add documentation for MT8365

Add documentation for MT8365 platform. Update supported boards list.
Update toolchains list supported in the Zephyr SDK.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

andrew-mtk commented:
@kartben, I have updated documentation and added info about MT8365

nordicjm commented:
CI is failing so you need to fix the issue causing that

nordicjm commented:
> > CI is failing so you need to fix the issue causing that
> 
> @nordicjm , to make twister not fail, [zephyrproject-rtos/hal_xtensa#36](https://github.com/zephyrproject-rtos/hal_xtensa/pull/36) has to be merged first. Than I can update `hal_xtensa` hash in west.xml pointing to `mt8365`.

where is the manifest update here? You need to add a manifest update to the pull request in west.yml and CI here **must** pass before that PR can be merged

andrew-mtk commented:
> where is the manifest update here? You need to add a manifest update to the pull request in west.yml and CI here **must** pass before that PR can be merged

@nordicjm , to add manifest update here I need to have updated hash from `hal_xtensa` repo that includes [https://github.com/zephyrproject-rtos/hal_xtensa/pull/36](https://github.com/zephyrproject-rtos/hal_xtensa/pull/36). Until that PR is merged I can't update west.yml over here.

nordicjm commented:
> > where is the manifest update here? You need to add a manifest update to the pull request in west.yml and CI here **must** pass before that PR can be merged
> 
> @nordicjm , to add manifest update here I need to have updated hash from `hal_xtensa` repo that includes [zephyrproject-rtos/hal_xtensa#36](https://github.com/zephyrproject-rtos/hal_xtensa/pull/36). Until that PR is merged I can't update west.yml over here.

Sure you can, see an example of it being done here: https://github.com/zephyrproject-rtos/zephyr/pull/98964/files#diff-d7f4082837dc3cb270fcf1abd68188216fdadee294e01611b8ca59ddea7fdbb3R213

andrew-mtk committed:
manifest: update hal_xtensa to add support for MT8365

Update hal_xtensa to add support for MT8365 platform.

Signed-off-by: Andrew Perepech <andrew.perepech@mediatek.com>

nordicjm commented on soc/mediatek/mt8xxx/CMakeLists.txt:
```diff
@@ -37,9 +40,9 @@ board_set_rimage_target(${CONFIG_SOC})
+  COMMAND west $<$<BOOL:${CMAKE_VERBOSE_MAKEFILE}>:--verbose> sign
                      --if-tool-available --tool rimage --build-dir ${CMAKE_BINARY_DIR}
-	DEPENDS ${CMAKE_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
+  DEPENDS ${CMAKE_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
```
```suggestion
  DEPENDS ${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME}
```

nordicjm commented on boards/mediatek/index.rst:
```diff
@@ -93,60 +95,11 @@ but this is still unintegrated.
-   export ZEPHYR_TOOLCHAIN_VARIANT=cross-compile
+The MT8195, MT818X and MT8365 toolchains are already part of the Zephyr
+SDK, so builds for the ``mt8195/mt8195/adsp``, ``mt8186/mt8186/adsp``,
+``mt8188/mt8188/adsp`` and ``mt8365/mt8365/adsp`` board should work out
```
```suggestion
``mt8188/mt8188/adsp`` and ``mt8365/mt8365/adsp`` board targets should work out
```

nordicjm commented on west.yml:
```diff
@@ -280,7 +280,7 @@ manifest:
         - hal
     - name: hal_xtensa
-      revision: 3cc9e3a9360be5c96c956dce84064b85439b6769
+      revision: pull/36/head
```
manifest update must come before things that need it (move this commit first)

nordicjm reviewed (changes requested):
Doc issue is https://github.com/zephyrproject-rtos/zephyr/issues/99031 which will be fixed on rebase. Noticed some other things with the changes files here

nordicjm commented on soc/mediatek/mt8xxx/CMakeLists.txt:
```diff
@@ -1,18 +1,21 @@
+  DEPENDS ${ZEPHYR_FINAL_EXECUTABLE}
+  COMMAND ${PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/gen_img.py
+    ${ZEPHYR_BINARY_DIR}/${KERNEL_ELF_NAME}
+    ${CMAKE_BINARY_DIR}/zephyr/zephyr.img)
```
yes, sorry

andrew-mtk commented:
> This is missing a documentation page

@kartben , @andyross , could you re-review the updated changes. The PR is in `Changes requested` state and I cant change it to `Review required`, though I have pushed the updates.

nordicjm commented:
Maintainer of xtensa repo needs to merge the PR there and then update the pull request link in the manifest here with the SHA

aescolar commented:
Re-triggering CI
@kartben could you check if your review comments have been addressed?
